pymc3/step_methods/hmc/base_hmc.py

```diff
@@ -53,10 +53,10 @@ def __init__(self, vars=None, scaling=None, step_scale=0.25, is_cov=False,
             `energy`, and `random` methods.
         **theano_kwargs: passed to theano functions
         """
-        model = modelcontext(model)
+        self.model = modelcontext(model)
```

Comment:
letting the model leak into objects causes trouble for various reasons - can you move this to the line after you check for finite energy?

Specifically:

```
if not np.isfinite(start.energy):
    model = modelcontext(model)
    ...
```

Reply:
Is it ok to still have self.model = modelcontext(model)? Or do I need to remove that completely?

Reply:
I'd remove it unless there is a good reason (dependencies get really messy when every object always has a pointer to the model context)

Reply:
Otherwise I get an unbounded local variable.

Reply:
From what?

Reply:
If I do model = modelcontext(model) without declaring a self.model. I get an unbounded local variable